internal/baseservice/base_service.go

```diff
@@ -106,6 +127,53 @@ func (s *BaseService) CancellableSleepRandomBetween(ctx context.Context, sleepDu
 	s.CancellableSleep(ctx, time.Duration(randutil.IntBetween(s.Rand, int(sleepDurationMin), int(sleepDurationMax))))
 }
 
+// CancellableSleepRandomBetween sleeps for a random duration between the given
+// bounds (max bound is exclusive), but returns early if context has been
+// cancelled. If the service's `DisableSleep` option is set (usually appropriate
+// in tests), sleep is skipped automatically.
+//
+// This variant returns a channel that should be waited on and which will be
+// closed when either the sleep or context is done.
+func (s *BaseService) CancellableSleepRandomBetweenC(ctx context.Context, sleepDurationMin, sleepDurationMax time.Duration) <-chan struct{} {
```

Comment:
Okay, so a couple of these new functions aren't used here yet, but gonna ask you to take some of it on faith. What happened is that I was working on this for a different branch, then started working on this project, and realized I wanted one of the functions for this branch. Instead of pulling in the one specific function I wanted, I just pulled in all the base service changes so I didn't have to tease apart all the tests I'd added and everything.